lightly/loss/byol_loss.py

```diff
@@ -0,0 +1,189 @@
+        super().__init__()
+        self.net = net
+        self.layer = layer
+
+        self.projector = None
+        self.projection_size = projection_size
+        self.projection_hidden_size = projection_hidden_size
+
+        self.hidden = None
+        self.hook_registered = False
+
+    def _find_layer(self):
+        if type(self.layer) == str:
+            modules = dict([*self.net.named_modules()])
+            return modules.get(self.layer, None)
+        elif type(self.layer) == int:
+            children = [*self.net.children()]
+            return children[self.layer]
+        return None
+
+    def _hook(self, _, __, output):
+        self.hidden = self.flatten(output)
+
+    def _register_hook(self):
+        layer = self._find_layer()
+        assert layer, f'hidden layer ({self.layer}) not found'
+        layer.register_forward_hook(self._hook)
+        self.hook_registered = True
+
+    @singleton('projector')
+    def get_projector(self, hidden):
+        _, dim = hidden.shape
+        projector = MLP(dim, self.projection_size, self.projection_hidden_size)
+        return projector.to(hidden)
+
+    def get_representation(self, x):
+        if self.layer == -1:
+            return self.net(x)
+
+        if not self.hook_registered:
+            self._register_hook()
+
+        _ = self.net(x)
+        hidden = self.hidden
+        self.hidden = None
+        assert hidden is not None, f'hidden layer {self.layer} never emitted an output'
+        return hidden
+
+    @staticmethod
+    def default(val, def_val):
+        return val if val else def_val
+
+    @staticmethod
+    def flatten(t):
+        return t.reshape(t.shape[0], -1)
+
+
+class OnlineModel:
+    """
+    Online student model, encapsulates an encoder and a predictor.
+    """
+
+    def __init__(self, net, projection_size, projection_hidden_size, hidden_layer):
+        self.encoder = NetWrapper(net, projection_size, projection_hidden_size, layer=hidden_layer)
+        self.predictor = MLP(projection_size, projection_size, projection_hidden_size)
+
+    def make_prediction(self, image):
+        return self.predictor(self.encoder(image))
+
+
+class TargetModel:
+    """
+    Mean Teacher Model that takes an exponential moving average weighed by beta of past observed weights.
+    """
+
+    def __init__(self, beta, encoder, ema):
+        self.beta = beta
+        self.encoder = encoder
+        self.ema = ema
+
+    def update_with_online_encoder(self, online_encoder):
+        for current_params, ma_params in zip(online_encoder.parameters(), self.encoder.parameters()):
+            old_weight, up_weight = ma_params.data, current_params.data
+            ma_params.data = self._update_average(old_weight, up_weight)
+
+    def _update_average(self, old, new):
+        if self.ema:
+            return old * self.beta + (1 - self.beta) * new
+        else:
+            return new
+
+
```

Comment:
The implementation of the loss should follow the style of the current implementations. I.e. the loss takes as input the outputs of the model. See:
- [Lightly at a Glance: Training](https://docs.lightly.ai/getting_started/lightly_at_a_glance.html)
- [NTXentLoss implementation](https://github.com/lightly-ai/lightly/blob/develop/lightly/loss/ntx_ent_loss.py)
- [Negative Cosine Similarity Loss by busycalibrating](https://github.com/lightly-ai/lightly/pull/61/files)

This allows for a more flexible setup where a single loss function can be combined with different architectures.